### PR TITLE
Testing wetzel automated reference output.  DO NOT MERGE

### PR DESCRIPTION
**Pull request by emackey**
## DO NOT MERGE

This is simply to show the diff of what would happen if we took automated [wetzel](https://github.com/CesiumGS/wetzel) output and blindly used it as a replacement for the reference section.

**NOTE:** If there are functional differences between this and what's in the reference section, it may indicate a disparity between the main reference section and the JSON schemas.

The command-line used to generate this was:

```
$ bin/wetzel.js -l2 -a=cqo ../glTF/specification/2.0/schema/glTF.schema.json > ../wetzel_output/ref_2020-10-28.md
```

### CODE REVIEW

**Comment by emackey** [comment]
There are some distinct improvements here, now that I'm looking in detail.  A number of `integer` references have been replaced by `glTFid` for example, and `object` or `any` replaced by explicit types.

Also, the sort order changed to take into account the hierarchy.  For example `orthographic` appears in the "Camera" section now, not alphabetized down by the O's.

The schema links have disappeared, but it looks like I'm missing a command-line arg for that.

**Comment by emackey** [comment]
Command line has changed:

```
$ bin/wetzel.js -l2 -a=cqo -p schema ../glTF/specification/2.0/schema/glTF.schema.json > ../wetzel_output/test2.md
```

**Comment by emackey** [comment]
Re: Sort order, and titles being different.  This isn't because of any change in wetzel!  It's because of 113a75b74442b0061a.

That small diff in the schema files produces huge diffs in the wetzel output.

**Comment by emackey** [comment]
Of course we wouldn't want to actually revert the schema title names, but I've done so on this branch, just for the purpose of making the wetzel output match with cleaner diffs.

File: specification/2.0/README.md [hunk 1]
```diff
@@ -2310,12 +2288,10 @@ Additional properties are allowed.
 
 Application-specific data.
 
-> **Implementation Note:** Although extras may have any type, it is common for applications to
-store and access custom data as key/value pairs. As best practice, extras should be an Object
-rather than a primitive value for best portability.
```
Comment (emackey) [inline specification/2.0/README.md:2315]: This `extras` implementation note is not in the schema, and is lost due to automatic regeneration.

File: specification/2.0/README.md [hunk 2]
```diff
@@ -2664,43 +2677,43 @@ The user-defined name of this object.  This is not necessarily unique, e.g., an
...
 
-A tangent space normal map. The texture contains RGB components in linear space. Each texel represents the XYZ components of a normal vector in tangent space. Red [0 to 255] maps to X [-1 to 1]. Green [0 to 255] maps to Y [-1 to 1]. Blue [128 to 255] maps to Z [1/255 to 1]. The normal vectors use OpenGL conventions where +X is right and +Y is up. +Z points toward the viewer. In GLSL, this vector would be unpacked like so: `vec3 normalVector = tex2D(normalMap, texCoord) * 2 - 1`. Client implementations should normalize the normal vectors before using them in lighting equations.
+A tangent space normal map. The texture contains RGB components in linear space. Each texel represents the XYZ components of a normal vector in tangent space. Red [0 to 255] maps to X [-1 to 1]. Green [0 to 255] maps to Y [-1 to 1]. Blue [128 to 255] maps to Z [1/255 to 1]. The normal vectors use OpenGL conventions where +X is right and +Y is up. +Z points toward the viewer. In GLSL, this vector would be unpacked like so: `float3 normalVector = tex2D(<sampled normal map texture value>, texCoord) * 2 - 1`. Client implementations should normalize the normal vectors before using them in lighting equations.
```
Comment (emackey) [inline specification/2.0/README.md:2700]: The formula is the same, but expressed with different keywords in schema vs. reference.

File: specification/2.0/README.md [hunk 3]
```diff
@@ -2931,59 +2944,59 @@ Reference to a texture.
...
-* **Type**: `integer`
+* **Type**: [`glTFid`](#reference-gltfid)
 * **Required**: Yes
 * **Minimum**: ` >= 0`
 
-#### normalTextureInfo.texCoord
+#### material.normalTextureInfo.texCoord
 
-This integer value is used to construct a string in the format TEXCOORD_<set index> which is a reference to a key in mesh.primitives.attributes (e.g. A value of 0 corresponds to TEXCOORD_0).
+This integer value is used to construct a string in the format `TEXCOORD_<set index>` which is a reference to a key in mesh.primitives.attributes (e.g. A value of `0` corresponds to `TEXCOORD_0`). Mesh must have corresponding texture coordinate attributes for the material to be applicable to it.
```
Comment (emackey) [inline specification/2.0/README.md:2967]: The last sentence was added to schema, but not added to reference.

File: specification/2.0/README.md [hunk 4]
```diff
@@ -2992,33 +3005,33 @@ Reference to a texture.
...
-* **Type**: `integer`
+* **Type**: [`glTFid`](#reference-gltfid)
 * **Required**: Yes
 * **Minimum**: ` >= 0`
 
-#### occlusionTextureInfo.texCoord
+#### material.occlusionTextureInfo.texCoord
 
-This integer value is used to construct a string in the format TEXCOORD_<set index> which is a reference to a key in mesh.primitives.attributes (e.g. A value of 0 corresponds to TEXCOORD_0).
+This integer value is used to construct a string in the format `TEXCOORD_<set index>` which is a reference to a key in mesh.primitives.attributes (e.g. A value of `0` corresponds to `TEXCOORD_0`). Mesh must have corresponding texture coordinate attributes for the material to be applicable to it.
```
Comment (emackey) [inline specification/2.0/README.md:3028]: Last sentence missing again.

File: specification/2.0/README.md [hunk 5]
```diff
@@ -3055,67 +3068,67 @@ An orthographic camera containing properties to create an orthographic projectio
 
 |   |Type|Description|Required|
 |---|----|-----------|--------|
-|**xmag**|`number`|The floating-point horizontal magnification of the view.| :white_check_mark: Yes|
-|**ymag**|`number`|The floating-point vertical magnification of the view.| :white_check_mark: Yes|
+|**xmag**|`number`|The floating-point horizontal magnification of the view. Must not be zero.| :white_check_mark: Yes|
+|**ymag**|`number`|The floating-point vertical magnification of the view. Must not be zero.| :white_check_mark: Yes|
```
Comment (emackey) [inline specification/2.0/README.md:3072]: "Must not be zero" was added to schema, but not ref.

File: specification/2.0/README.md [hunk 6]
```diff
@@ -2086,14 +2023,14 @@ A view into a buffer generally representing a subset of the buffer.
 
 |   |Type|Description|Required|
 |---|----|-----------|--------|
-|**buffer**|`integer`|The index of the buffer.| :white_check_mark: Yes|
+|**buffer**|[`glTFid`](#reference-gltfid)|The index of the buffer.| :white_check_mark: Yes|
 |**byteOffset**|`integer`|The offset into the buffer in bytes.|No, default: `0`|
-|**byteLength**|`integer`|The length of the bufferView in bytes.| :white_check_mark: Yes|
+|**byteLength**|`integer`|The total byte length of the buffer view.| :white_check_mark: Yes|
```
Comment (emackey) [inline specification/2.0/README.md:2028]: I think the reference has better wording than the schema here, we should copy it down.

File: specification/2.0/README.md [hunk 7]
```diff
@@ -1869,80 +1871,15 @@ The user-defined name of this object.  This is not necessarily unique, e.g., an
...
-#### animation sampler.interpolation
-
-Interpolation algorithm.
-
-* **Type**: `string`
-* **Required**: No, default: `"LINEAR"`
-* **Allowed values**:
-   * `"LINEAR"` The animated values are linearly interpolated between keyframes. When targeting a rotation, spherical linear interpolation (slerp) should be used to interpolate quaternions. The number output of elements must equal the number of input elements.
-   * `"STEP"` The animated values remain constant to the output of the first keyframe, until the next keyframe. The number of output elements must equal the number of input elements.
-   * `"CUBICSPLINE"` The animation's interpolation is computed using a cubic spline with specified tangents. The number of output elements must equal three times the number of input elements. For each input element, the output stores three elements, an in-tangent, a spline vertex, and an out-tangent. There must be at least two keyframes when using this interpolation.
```
Comment (emackey) [inline specification/2.0/README.md:1923]: Missing from converted output?  May be the result of name changes or conflicts.

File: specification/2.0/README.md [hunk 8]
```diff
@@ -1869,80 +1871,15 @@ The user-defined name of this object.  This is not necessarily unique, e.g., an
...
-|**extensions**|`object`|Dictionary object with extension-specific objects.|No|
-|**extras**|`any`|Application-specific data.|No|
-
-Additional properties are allowed.
-
-* **JSON schema**: [animation.sampler.schema.json](schema/animation.sampler.schema.json)
-
-#### animation sampler.input :white_check_mark: 
-
-The index of an accessor containing keyframe input values, e.g., time. That accessor must have componentType `FLOAT`. The values represent time in seconds with `time[0] >= 0.0`, and strictly increasing values, i.e., `time[n + 1] > time[n]`.
```
Comment (emackey) [inline specification/2.0/README.md:1908]: Missing from converted output?  May be the result of name changes or conflicts.

File: specification/2.0/README.md [hunk 9]
```diff
@@ -1869,80 +1871,15 @@ The user-defined name of this object.  This is not necessarily unique, e.g., an
...
-   * `"STEP"` The animated values remain constant to the output of the first keyframe, until the next keyframe. The number of output elements must equal the number of input elements.
-   * `"CUBICSPLINE"` The animation's interpolation is computed using a cubic spline with specified tangents. The number of output elements must equal three times the number of input elements. For each input element, the output stores three elements, an in-tangent, a spline vertex, and an out-tangent. There must be at least two keyframes when using this interpolation.
-
-#### animation sampler.output :white_check_mark: 
-
-The index of an accessor containing keyframe output values. When targeting translation or scale paths, the `accessor.componentType` of the output values must be `FLOAT`. When targeting rotation or morph weights, the `accessor.componentType` of the output values must be `FLOAT` or normalized integer. For weights, each output element stores `SCALAR` values with a count equal to the number of morph targets.
```
Comment (emackey) [inline specification/2.0/README.md:1927]: Missing from converted output?  May be the result of name changes or conflicts.

File: specification/2.0/README.md [hunk 10]
```diff
@@ -2026,10 +1963,10 @@ A buffer points to binary geometry, animation, or skins.
 |   |Type|Description|Required|
 |---|----|-----------|--------|
 |**uri**|`string`|The uri of the buffer.|No|
-|**byteLength**|`integer`|The total byte length of the buffer view.| :white_check_mark: Yes|
```
Comment (emackey) [inline specification/2.0/README.md:2029]: This is worded better in the schema, and even contains a fix: This is a buffer, not a buffer view.

**Review by emackey — COMMENTED** [review]
Overall, with the sorting issues mostly removed, this isn't too bad a diff.

**Comment by donmccurdy** [comment]
Thanks for doing this, @emackey! I haven't done a full read-through yet, but other than a couple "Missing from converted output" issues, it seems like things haven't diverged as far as we thought they may have? And with a careful review, we could probably bring things in sync? Or do you see any bigger issues here?

If we plan to keep running wetzel after we finish the AsciiDoc conversion, perhaps we could have GitHub do some kind of lint step that verifies a re-run of wetzel does not change the spec, and flags PRs if they would bring us out of sync?

**Comment by javagl** [comment]
Indeed, thanks @emackey ! 

As a cross-reference: Note that some of the changes may vaguely be related to https://github.com/CesiumGS/wetzel/pull/33#issue-249722583 , and some other reasons for differences are in changes to the *schema*, as mentioned there.

There had been some considerations about updating the actual specs with the latest wetzel output, but I hesitated to do this, mainly due to the implications that this may have for external links that lead to `#anchors` in the README that may no longer be present afterwards...

But admittedly: All this has been long ago, so I'd need a refresher - maybe I can have a closer look at this PR in the next days.

File: specification/2.0/README.md (hunk 7 above)
Comment (emackey) [reply specification/2.0/README.md]: This was just a name conflict between animation samplers and texture samplers.

**Comment by emackey** [comment]
Thanks @donmccurdy and @javagl.  My plan is to open a separate PR with some spec cleanup once this is fully resolved (soon).

I do think it's worth having some automated process generate this section of the new AsciiDoctor version, going through Markdown as an intermediate step, and using an include directive.  This has the benefit of making the JSON schemas a single source of truth on the reference section, and the include directive will prevent there being any separate file to get out of sync.

Note to self:  Command line changed again.  The "ignore" directive expects lower-case input, only.

```
$ bin/wetzel.js -l2 -a=cqo -p schema -i '["gltfchildofrootproperty.schema.json", "gltfid.schema.json", "gltfproperty.schema.json"]' ../glTF/specification/2.0/schema/glTF.schema.json > ../output.md
```

File: specification/2.0/README.md [hunk 11]
```diff
@@ -2326,163 +2324,163 @@ The root object for a glTF asset.
...
-|**bufferViews**|bufferView `[1-*]`|An array of bufferViews.|No|
-|**cameras**|camera `[1-*]`|An array of cameras.|No|
-|**images**|image `[1-*]`|An array of images.|No|
-|**materials**|material `[1-*]`|An array of materials.|No|
-|**meshes**|mesh `[1-*]`|An array of meshes.|No|
-|**nodes**|node `[1-*]`|An array of nodes.|No|
-|**samplers**|sampler `[1-*]`|An array of samplers.|No|
+|**accessors**|[`accessor`](#reference-accessor) `[1-*]`|An array of accessors.|No|
+|**animations**|[`animation`](#reference-animation) `[1-*]`|An array of keyframe animations.|No|
+|**asset**|[`asset`](#reference-asset)|Metadata about the glTF asset.|No|
```
Comment (emackey) [inline specification/2.0/README.md:2329]: Whoops, somehow "required" changed to "No" for asset.  Might be a Wetzel bug?

Comment (emackey) [reply specification/2.0/README.md]: Fixed in https://github.com/CesiumGS/wetzel/pull/38

**Comment by emackey** [comment]
Looking carefully at these diffs, I think all hand-edits have been resolved.  There's one paragraph break manually inserted in the reference section, but it's not critical and we don't have a way to put a paragraph break into a glTF schema description.

File: specification/2.0/README.md (hunk 1 above)
Comment (emackey) [reply specification/2.0/README.md:2315]: This was added in https://github.com/CesiumGS/wetzel/pull/42.

**Pull request closed without merging**